### PR TITLE
OGD-143 implemented sandbox employments endpoint

### PR DESCRIPTION
**Pull request by edinhodzic**
The only thing missing from this is the requirement to have an optional and default `toDate` parameter. Since I'll have to get my head around the `QueryStringBindable` implementation we have and since that will also affect the income endpoint, I thought I'd do a pull request at this stage while I work on delivering that requirement.

### CODE REVIEW

File: app/uk/gov/hmrc/individualsapi/controllers/EmploymentsController.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import play.api.hal.HalLink
+import play.api.libs.json.Json.{obj, toJson}
+import play.api.mvc.Action
+import play.api.mvc.hal._
+import uk.gov.hmrc.individualsapi.models.JsonFormatters.employmentJsonFormat
+import uk.gov.hmrc.individualsapi.services.{EmploymentsService, SandboxEmploymentsService}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class EmploymentsController(employmentsService: EmploymentsService) extends CommonController {
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/controllers/EmploymentsController.scala]: abstract class

Comment (edinhodzic) [inline app/uk/gov/hmrc/individualsapi/controllers/EmploymentsController.scala]: This class is [instantiated in a test](https://github.com/hmrc/individuals-api/blob/062d6da6f85de53bbeba46df6aaa3aa1e3eaadf1/test/unit/uk/gov/hmrc/individualsapi/controllers/EmploymentsControllerSpec.scala#L43). Making it abstract would mean defining a new implementation of it for test purposes. Shall we still make it abstract?

Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/controllers/EmploymentsController.scala]: I would say so.
The tests should reflect the code not the opposite :-)

File: app/uk/gov/hmrc/individualsapi/models/Individual.scala [hunk 2]
```diff
@@ -27,33 +28,70 @@ case class Individual(matchId: String,
                       employments: Seq[Employment],
                       income: Seq[Income])
 
-case class Employment(employerPayeReference: String, employerName: String, startDate: LocalDate, endDate: LocalDate)
+case class Address(line1: String, line2: String, line3: String, line4: Option[String], line5: Option[String], line6: Option[String], postcode: String)
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: only line1 and postcode are mandatory. Others are optional

File: app/uk/gov/hmrc/individualsapi/services/EmploymentsService.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.joda.time.Interval
+import uk.gov.hmrc.individualsapi.models
+import uk.gov.hmrc.individualsapi.models.{Employment, MatchNotFoundException}
+
+import scala.concurrent.Future
+import scala.concurrent.Future.{failed, successful}
+
+trait EmploymentsService {
+  def paye(matchId: String, interval: Interval): Future[Seq[Employment]] = ???
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/services/EmploymentsService.scala]: remove = ???

**Comment by edinhodzic** [comment]
I've now added the default value for `toDate` when a `fromDate` is present but a `toDate` is omitted.

File: app/uk/gov/hmrc/individualsapi/models/Individual.scala [hunk 4]
```diff
@@ -27,33 +28,70 @@ case class Individual(matchId: String,
...
+    val employmentInterval = new Interval(startTime, endTime)
+    Dates.overlap(interval, employmentInterval)
+  }
+
+}
+
+case class Payment(date: LocalDate, amount: Double) {
+
+  def isPaidWithin(interval: Interval): Boolean =
+    interval.contains(date.toDateTimeAtCurrentTime)
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: the toDateTimeAtCurrentTime worries me.
What happens when interval is from 01/01/17 to 02/02/17 and date is 02/02/17 with current time being like 23.00.
Is it in the interval?

Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: can we have some unit test for those methods isPaidWithin, overlap, containsPaymentWithin?

**Comment by edinhodzic** [comment]
With the latest commits, the main change is in `IntervalQueryStringBinder`. This is responsible for deriving an interval based on `fromDate` and `toDate` parameters. Whereas before it disregarded the time of day and thereby defaulted to `00:00:00.000` for both `fromDate` and `toDate`, it now ensures that the time of day for `toDate` is instead `23:59:59.999`. This effectively means that before `toDate` was exclusive and now it is inclusive.

File: app/uk/gov/hmrc/individualsapi/models/Individual.scala [hunk 5]
```diff
@@ -27,33 +27,71 @@ case class Individual(matchId: String,
...
                       income: Seq[Income])
 
-case class Employment(employerPayeReference: String, employerName: String, startDate: LocalDate, endDate: LocalDate)
+case class Address(line1: String, line2: Option[String], line3: Option[String], line4: Option[String], line5: Option[String], line6: Option[String], postcode: String)
 
-case class Payment(date: LocalDate, amount: Double)
+case class Employer(payeReference: String, name: String, address: Address)
 
-case class Income(employerPayeReference: String, payments: Seq[Payment])
+case class Employment(startDate: LocalDate, endDate: LocalDate, employer: Employer, payFrequency: String)
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: endDate: Option[LocalDate]

Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: payFrequency: Option[String]

File: app/uk/gov/hmrc/individualsapi/models/Individual.scala [hunk 6]
```diff
@@ -27,33 +27,71 @@ case class Individual(matchId: String,
...
                       income: Seq[Income])
 
-case class Employment(employerPayeReference: String, employerName: String, startDate: LocalDate, endDate: LocalDate)
+case class Address(line1: String, line2: Option[String], line3: Option[String], line4: Option[String], line5: Option[String], line6: Option[String], postcode: String)
 
-case class Payment(date: LocalDate, amount: Double)
+case class Employer(payeReference: String, name: String, address: Address)
 
-case class Income(employerPayeReference: String, payments: Seq[Payment])
+case class Employment(startDate: LocalDate, endDate: Option[LocalDate], employer: Employer, payFrequency: Option[String])
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: Following DES schema, we also need to change:
startDate: Option[LocalDate]
name: Option[String]
address: Option[String]

File: app/uk/gov/hmrc/individualsapi/models/Individual.scala [hunk 7]
```diff
@@ -27,33 +27,71 @@ case class Individual(matchId: String,
...
+
+object Employment {
+
+  def overlap(interval: Interval)(employment: Employment): Boolean = {
+    val oneMillisecondBeforeMidnight = LocalTime.MIDNIGHT.minusMillis(1)
+    val startTime = employment.startDate.toDate.getTime
+    val endTime = employment.endDate.getOrElse(LocalDate.now).toDateTime(oneMillisecondBeforeMidnight).getMillis
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: as we do not care about hours, I feel we should just have toDateTimeAtStartOfDay everywhere (here and in the binder.)

Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/models/Individual.scala]: Can we have a helper like
  protected def toInterval(fromDate: LocalDate, toDate: LocalDate): Interval =
    new Interval(fromDate.toDateTimeAtStartOfDay.getTime, toDate.toDateTimeAtStartOfDay.getTime)
and use it here and in the binder?

File: app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.joda.time.{Interval, LocalDate, LocalTime}
+import uk.gov.hmrc.individualsapi.Binders.localDateQueryStringBinder
+
+class IntervalQueryStringBinder extends AbstractQueryStringBindable[Interval] {
+
+  private val oneMillisecondBeforeMidnight = LocalTime.MIDNIGHT.minusMillis(1)
+
+  private def interval(fromDate: LocalDate, toDate: LocalDate): Either[String, Interval] = try {
+    Right(new Interval(fromDate.toDateTimeAtStartOfDay, toDate.toDateTime(oneMillisecondBeforeMidnight)))
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala]: same as before. I think it is easier to have toDateTimeAtStartOfDay everywhere

**Comment by tom59** [comment]
missing app.routes + RAML

File: app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  } catch {
+    case _: Throwable => Left(errorResponse("Invalid time period requested"))
+  }
+
+  override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, Interval]] = {
+    for {
```
Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala]: there is no really need for localDateQueryStringBinder anymore. we should remove it?
That would simplify the code below to something like:
    (params.get("fromDate").flatMap(_.headOption), params.get("toDate").flatMap(_.headOption)) match {
      case (Some(fromDate), Some(toDate)) =>
        Right(new Interval(from, to))
...

**Comment by edinhodzic** [comment]
> missing app.routes + RAML

@tom59 this is a sandbox implementation so it's in `sandbox.routes`.

As for RAML I was expecting a ticket like one of these:

- https://jira.tools.tax.service.gov.uk/browse/OGD-208
- https://jira.tools.tax.service.gov.uk/browse/OGD-209
- https://jira.tools.tax.service.gov.uk/browse/OGD-237
- https://jira.tools.tax.service.gov.uk/browse/OGD-250
- https://jira.tools.tax.service.gov.uk/browse/OGD-267

Or do we want to on this occasion do it in one ticket?

File: app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala (hunk 9 above)
Comment (edinhodzic) [inline app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala]: This isn't necessarily new code as I [extracted these binders into here from anonymous classes](https://github.com/hmrc/individuals-api/pull/41/commits/d4a955c06d4ea37a45a3bbf15083256db29216a9). In any case, the `LocalDateQueryStringBinder` class not only get the parameters but also does some validation. If only in the interest of separation of concerns, I think they should remain separate. I mean if we were to refactor it we'd have to move that code into here and I don't think we could make reduce the amount of code or make it neater. But I could be wrong. Feel free to give it a go see what we end up with.

Comment (tom59) [inline app/uk/gov/hmrc/individualsapi/util/IntervalQueryStringBinder.scala]: ok i will try on monday

**Comment by tom59** [comment]
@edinhodzic I have done a commit to remove the date binder. If you are happy with it, we can merge it.

**Comment by edinhodzic** [comment]
@tom59 looks good to me.

**Pull request merged**